### DISCUSSION TITLE
Lesson 10: "Setup failed: Custom Error" - not showing the custom error name

### DISCUSSION THREAD

**Discussion by DanielSlavov** [Cyfrin/foundry-full-course-cu#721] [Q&A]
Hello everybody! When setting up the OurTokenTest, Patrick at first doesn't set the vm.prank correctly and gets "**Setup failed: ERC20: transfer amount exceeds balance**", on the other hand when I on my side don't set the vm.prank correctly and run forge test I get "**Setup failed: Custom Error e450d38c:**". If you have an idea why is this happening and why  **I don't get the correct error message** or maybe if Im missing something, please share. Obviously the error comes from the OpenZeppelin's ERC20 contract, but the error's name doesn't get displayed on my console. Like in the course I have extended the latest OpenZeppelin ERC20 contract imported with foundry from github and I have tried setting solc to later version but no success. Getting an error and not knowing its name or cause is a bummer...Thanks in advance!!!

**Comment by DanielSlavov** [comment]
So I found out that the tutorial's older version of the openzeppelin's ERC20 contract was using reqiure with message and 
this leads to the correct error message in the tutorial. The most recent version of the contract uses revert with a Custom Error. This though doesn't solve my problem and I am still unable to find how to display the Custom Error name when it is thrown when running forge test...

**Reply by theirrationalone** [reply]
Put custom errors inside at the top of contract. And try this... 


```solidity
vm.startPrank(USER);
vm.expectRevert(ContractName.ContractName__AnyError.selector);    // Or after `ContractName.` your custom error. 
```

Or if you have parameters in your code... 

```solidity
vm.startPrank(USER);
vm.expectRevert(abi.encodeWithSelector(ContractName.ContractName__AnyError.selector, parameter1, parameter2, ...);    // Or after `ContractName.` your custom error. 
```

You can get a `USER` by initializing it at the top of your test script like this.... 

```solidity
import {Test} from "forge-std/Test.sol";

contract ContractTest is Test{
    address USER = makeAddr("user") ;
    
    function setUp() external {
        // statements... 
    }

    function testYourTest() public {
        // things you wanna test... 
    }
```

remember, `setUp` function will be executed once for each test... Means... 

`setUp` -> test1 done  ✅
`setUp` -> test2 done  ✅
`setUp` -> test3 done  ✅
. 
. 
. 
`setUp` -> testN done ✅

**Comment by DanielSlavov** [comment]
Sorry if I have mislead you and if I haven't understood correctly, but I think this answer does not answer my question. My question is why does the ERC20's **ERC20InsufficientBalance** error appear as **Custom Error e450d38c** in foundry console and not with its original name.  **The problem is not that the error is occurring but that it's name is hidden**. This scenario has nothing to do with testing or deploying nor with pranking or expecting a revert. Simply asked why is the error's original name not showed in the console and "Custom Error" is showed instead. To be clear again this is a simple runtime error due to Insufficient Balance, **I am not trying to** cause it intentionally and to expect a revert after that in order to test it.  Thanks!

**Comment by theirrationalone** [comment]
`e450d38c` is the selector encoded value. Playing with strings is such an expensive game so that's why we make use selector provided by error generator.

**Reply by DanielSlavov** [reply]
Is there a way to actually decode the selector encoded value?

**Reply by theirrationalone** [reply]
Try this out... [abi.decode](https://coinsbench.com/solidity-abi-encode-and-decode-b339eb52c5b5)

**Comment by farman1094** [comment]
I was also getting this same issue. That custom error is reverting in `Sighash`. I also have to check through error by error by converting into the `Sighash` first. Then comparing into the actual error to find out that's the error which is reverting.

**Reply by EngrPips** [reply]
Interesting.
